### PR TITLE
Handle statics in MIR as const pointers

### PR DESCRIPTION
**Pull request by matthewjasper**
This is the first PR towards the goal of removing `PlaceBase::Static`. In this PR:

* Statics are lowered to dereferencing a const pointer.
* The temporaries holding such pointers are tracked in MIR, for the most part this is only used for diagnostics. There are two exceptions:
    * The borrow checker has some checks for thread-locals that directly use this data.
    * Const checking will suppress "cannot dereference raw pointer" diagnostics for pointers to `static mut`/`extern static`. This is to maintain the current behaviour (12 tests fail otherwise).

The following are left to future PRs (I think that @spastorino will be working on the first 3):

* Applying the same treatments to promoted statics.
* Removing `PlaceBase::Static`.
* Replacing `PlaceBase` with `Local`.
* Moving the ever growing collection of metadata that we have for diagnostics in MIR passes somewhere more appropriate.

r? @oli-obk

### CODE REVIEW

File: src/librustc_mir/build/expr/as_temp.rs [hunk 1]
```diff
@@ -63,6 +64,11 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
             if let Some(tail_info) = this.block_context.currently_in_block_tail() {
                 local_decl = local_decl.block_tail(tail_info);
             }
+            if let ExprKind::StaticRef { def_id, .. } = expr.kind {
+                let attrs = this.hir.tcx().get_attrs(def_id);
+                let is_thread_local = attrs.iter().any(|attr| attr.check_name(sym::thread_local));
```
Comment (Centril) [inline src/librustc_mir/build/expr/as_temp.rs]:
```suggestion
                let is_thread_local = this.hir.tcx().has_attr(def_id, sym::thread_local);
```

File: src/librustc_mir/build/expr/as_constant.rs [hunk 2]
```diff
@@ -45,6 +45,13 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
                     literal,
                 }
             },
+            ExprKind::StaticRef { literal, .. } => {
```
Comment (oli-obk) [inline src/librustc_mir/build/expr/as_constant.rs]: This variant should have a bit of documentation explaining that it is equivalent to `ExprKind::Constant´ except that it registers some diagnostic info

**Review by oli-obk — APPROVED** [review]
This is great, a much more principled approach than what I was attempting

**Comment by spastorino** [comment]
@matthewjasper great work, indeed I'm working on the first 3.

File: src/librustc/mir/mod.rs [hunk 3]
```diff
@@ -737,7 +737,8 @@ pub struct LocalDecl<'tcx> {
     /// therefore it need not be visible across crates. pnkfelix
     /// currently hypothesized we *need* to wrap this in a
     /// `ClearCrossCrate` as long as it carries as `HirId`.
```
Comment (spastorino) [inline src/librustc/mir/mod.rs]: Does this comment  about `ClearCrossCrate` above still applies?

Comment (spastorino) [reply src/librustc/mir/mod.rs]: Maybe you want to move that into `LocalInfo`?

File: src/librustc_mir/build/expr/as_constant.rs (hunk 2 above)
Comment (matthewjasper) [reply src/librustc_mir/build/expr/as_constant.rs]: I've added this to the doc comment on the variant.

**Review by spastorino — APPROVED** [review]

**Comment by spastorino** [comment]
IMO this is ready to r+, would leave this up to @oli-obk though

**Comment by bjorn3** [comment]
Before this PR is was possible to always force a `Operand::Constant` to a const allocation and codegen a reference to that alloc. After this PR that would make TLS statics impossible to handle, as you would get reference to a reference to the static. You would then not know that the second reference should be treated as TLS access.

**Comment by oli-obk** [comment]
Do pointers track the TLS-ness of what they point to? If so, we can make const alloc codegen add those markers

**Comment by bjorn3** [comment]
No. Only when an alloc gets directly referenced from a function, does the `cranelift_codegen::ir::GlobalValue` pointing to it get marked as TLS value. When creating a pointer to it using `global_value` cranelift then calls the necessary function to get the right pointer for the current thread.

I have been able to solve the problem for cg_clif in https://github.com/bjorn3/rustc_codegen_cranelift/commit/30de2a0351f44e95867fdc2dad48beb20ebe3b47. That commit makes the logic more like cg_llvm.

**Comment by oli-obk** [comment]
So... the status is you solved the problem, but we could still improve on it by making this part of codegen_ssa to duplicate less code?

**Comment by bjorn3** [comment]
> the status is you solved the problem

Yes

> but we could still improve on it by making this part of codegen_ssa to duplicate less code?

cg_ssa already does this, but I can't use it much in cg_clif yet, as it depends to much on several llvm properties.

**Pull request merged**